source/simulator/assemblers/newton_stokes.cc

```diff
@@ -148,6 +148,11 @@ namespace aspect
                                    :
                                    1;
 
+              // pre-compute the tensor contractions
+              std::vector<SymmetricTensor<2,dim> > tensor_contraction(stokes_dofs_per_cell);
+              for (unsigned int i = 0; i < stokes_dofs_per_cell; ++i)
+                tensor_contraction[i] = (viscosity_derivative_wrt_strain_rate * scratch.grads_phi_u[i]) * strain_rate;
+
               // symmetrize when the stabilization is symmetric or SPD
               if ((preconditioner_stabilization & Newton::Parameters::Stabilization::symmetric) != Newton::Parameters::Stabilization::none)
                 {
```

Comment:
don't you actually only need these tensor contraction terms if you go into this `if` block here? If so, maybe move the declaration and initialization above after the `if` here.